v2/services/authorize/client_retry_test.go

```diff
@@ -0,0 +1,345 @@
+	require.NoError(t, err)
+
+	clientDataStore, err := generateDatastore(ca, privateKey, caCertPEM, 3*24*time.Hour)
+	require.NoError(t, err)
+
+	tlsCredentials, err := loadTLSCredentials(serverDataStore)
+	require.NoError(t, err)
+
+	server, err := launchServer(tlsCredentials)
+	require.NoError(t, err)
+
+	c := authorize.CreateClient()
+	c.SetRequestTimeout(time.Millisecond)
+
+	err = c.DialUsingCredentialsManager(context.Background(), &mockCredentialsFetcher{ds: clientDataStore}, "localhost", "10000", "")
+	require.NoError(t, err)
+
+	server.Shutdown()
+
+	_, err = c.GetResourceWithContext(context.Background(), "", "")
+
+	require.EqualError(t, err, "rpc error: code = DeadlineExceeded desc = context deadline exceeded")
+}
+
+func TestReconnect(t *testing.T) {
+	privateKey, err := parseRSAKey()
+	require.NoError(t, err)
+
+	caCertPEM, err := generateCA(privateKey)
+	require.NoError(t, err)
+
+	serverDataStore, err := generateDatastore(ca, privateKey, caCertPEM, 10*365*24*time.Hour)
+	require.NoError(t, err)
+
+	clientDataStore, err := generateDatastore(ca, privateKey, caCertPEM, 24*time.Hour)
+	require.NoError(t, err)
+
+	tlsCredentials, err := loadTLSCredentials(serverDataStore)
+	require.NoError(t, err)
+
+	server, err := launchServer(tlsCredentials)
+	require.NoError(t, err)
+
+	c := authorize.CreateClient()
+
+	err = c.DialUsingCredentialsManager(context.Background(), &mockCredentialsFetcher{ds: clientDataStore}, "localhost", "10000", "")
+	require.NoError(t, err)
+
+	server.Shutdown()
+
+	server, err = launchServer(tlsCredentials)
+	require.NoError(t, err)
+	defer server.Shutdown()
+
+	_, err = c.GetResourceWithContext(context.Background(), "", "")
+
+	require.NoError(t, err)
+}
+
+func TestRetryPolicy(t *testing.T) {
+	privateKey, err := parseRSAKey()
+	require.NoError(t, err)
+
+	caCertPEM, err := generateCA(privateKey)
+	require.NoError(t, err)
+
+	serverDataStore, err := generateDatastore(ca, privateKey, caCertPEM, 10*365*24*time.Hour)
+	require.NoError(t, err)
+
+	clientDataStore, err := generateDatastore(ca, privateKey, caCertPEM, 3*24*time.Hour)
+	require.NoError(t, err)
+
+	tlsCredentials, err := loadTLSCredentials(serverDataStore)
+	require.NoError(t, err)
+
+	server, err := launchServer(tlsCredentials)
+	require.NoError(t, err)
+	defer server.Shutdown()
+
+	c := authorize.CreateClient()
+
+	err = c.DialUsingCredentialsManager(context.Background(), &mockCredentialsFetcher{ds: clientDataStore}, "localhost", "10000", "")
+	require.NoError(t, err)
+
+	err = c.AddResourceWithContext(context.Background(), common.Origin{
+		Id:       "",
+		Type:     "",
+		Provider: "",
+	})
+	require.NoError(t, err)
+}
+
+func TestClientHandshake_CertificateAboutToExpire(t *testing.T) {
+	privateKey, err := parseRSAKey()
+	require.NoError(t, err)
+
+	caCertPEM, err := generateCA(privateKey)
+	require.NoError(t, err)
+
+	ds, err := generateDatastore(ca, privateKey, caCertPEM, 24*time.Hour-time.Second)
```

Comment:
Maybe we should store the grace period in a well-named constant? It would make the code and the test more readable